### PR TITLE
`sage.rings.polynomial`: Modularization fixes, `# needs`

### PR DESCRIPTION
**Pull request by mkoeppe**
- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/rings/lazy_series.py [hunk 1]
```diff
@@ -5511,54 +5510,54 @@ def __call__(self, *args, check=True):
         EXAMPLES::
 
             sage: P.<q> = QQ[]
-            sage: s = SymmetricFunctions(P).s()
-            sage: L = LazySymmetricFunctions(s)
-            sage: f = s[2]
-            sage: g = s[3]
-            sage: L(f)(L(g)) - L(f(g))
+            sage: s = SymmetricFunctions(P).s()                                         # optional - sage.modules
```
Comment (dcoudert) [inline src/sage/rings/lazy_series.py]: block scope

File: src/sage/rings/lazy_series.py [hunk 2]
```diff
@@ -5511,54 +5510,54 @@ def __call__(self, *args, check=True):
...
 
-            sage: s = SymmetricFunctions(QQ).s()
-            sage: X = tensor([s[1],s[[]]])
-            sage: Y = tensor([s[[]],s[1]])
-            sage: S = LazySymmetricFunctions(s)
-            sage: S2 = LazySymmetricFunctions(tensor([s, s]))
-            sage: A = S(s[1,1,1])
-            sage: B = S2(X+Y)
-            sage: A(B)
+            sage: s = SymmetricFunctions(QQ).s()                                        # optional - sage.modules
```
Comment (dcoudert) [inline src/sage/rings/lazy_series.py]: block scope

File: src/sage/rings/abc.pyx [hunk 3]
```diff
@@ -177,7 +177,7 @@ cdef class RealField(Field):
     EXAMPLES::
 
         sage: import sage.rings.abc
-        sage: isinstance(RR, sage.rings.abc.RealField)
+        sage: isinstance(RR, sage.rings.abc.RealField)                                  # optional - sage.rings.real_mpfr
```
Comment (dcoudert) [inline src/sage/rings/abc.pyx]: `# needs` ?

File: src/sage/rings/polynomial/laurent_polynomial_ring.py [hunk 4]
```diff
@@ -480,23 +483,24 @@ def _element_constructor_(self, x):
 
             sage: U = LaurentPolynomialRing(QQ, 'a')
             sage: V = LaurentPolynomialRing(QQ, 'c')
-            sage: L.<a, b, c, d> = LaurentPolynomialRing(QQ)
-            sage: M = LaurentPolynomialRing(QQ, 'c, d')                                 # optional - sage.modules
-            sage: Mc, Md = M.gens()                                                     # optional - sage.modules
-            sage: N = LaurentPolynomialRing(M, 'a, b')                                  # optional - sage.modules
-            sage: Na, Nb = N.gens()                                                     # optional - sage.modules
-            sage: U(Na)                                                                 # optional - sage.modules
+            sage: L.<a, b, c, d> = LaurentPolynomialRing(QQ)                            # needs sage.modules
```
Comment (dcoudert) [inline src/sage/rings/polynomial/laurent_polynomial_ring.py]: block scope

File: src/sage/rings/polynomial/multi_polynomial_element.py [hunk 5]
```diff
@@ -1026,42 +1041,43 @@ def global_height(self, prec=None):
...
             6.43775164973640
 
         ::
 
             sage: R.<x> = QQ[]
-            sage: K.<k> = NumberField(x^2 + 1)                                          # optional - sage.rings.number_field
-            sage: Q.<q,r> = PolynomialRing(K, implementation='generic')                 # optional - sage.rings.number_field
-            sage: f = 12 * q                                                            # optional - sage.rings.number_field
-            sage: f.global_height()                                                     # optional - sage.rings.number_field
+            sage: K.<k> = NumberField(x^2 + 1)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_element.py]: block scope

File: src/sage/rings/polynomial/multi_polynomial_element.py [hunk 6]
```diff
@@ -1106,34 +1122,33 @@ def local_height(self, v, prec=None):
...
 
         ::
 
             sage: R.<x> = QQ[]
-            sage: K.<k> = NumberField(x^2 - 5)                                          # optional - sage.rings.number_field
-            sage: T.<t,w> = PolynomialRing(K, implementation='generic')                 # optional - sage.rings.number_field
-            sage: I = K.ideal(3)                                                        # optional - sage.rings.number_field
-            sage: f = 1/3*t*w + 3                                                       # optional - sage.rings.number_field
-            sage: f.local_height(I)                                                     # optional - sage.rings.number_field sage.symbolic
+            sage: K.<k> = NumberField(x^2 - 5)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_element.py]: block scope except for last line needing `sage.symbolic` ?

File: src/sage/rings/polynomial/multi_polynomial_element.py [hunk 7]
```diff
@@ -1156,33 +1171,32 @@ def local_height_arch(self, i, prec=None):
...
             5.34710753071747
 
         ::
 
             sage: R.<x> = QQ[]
-            sage: K.<k> = NumberField(x^2 - 5)                                          # optional - sage.rings.number_field
-            sage: T.<t,w> = PolynomialRing(K, implementation='generic')                 # optional - sage.rings.number_field
-            sage: f = 1/2*t*w + 3                                                       # optional - sage.rings.number_field
-            sage: f.local_height_arch(1, prec=52)                                       # optional - sage.rings.number_field
+            sage: K.<k> = NumberField(x^2 - 5)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_element.py]: block scope

File: src/sage/rings/polynomial/multi_polynomial_ideal.py [hunk 8]
```diff
@@ -4646,10 +4650,10 @@ def groebner_basis(self, algorithm='', deg_bound=None, mult_bound=None, prot=Fal
         if not algorithm:
             try:
                 gb = self._groebner_basis_libsingular("groebner", deg_bound=deg_bound, mult_bound=mult_bound, *args, **kwds)
-            except (TypeError, NameError): # conversion to Singular not supported
+            except (TypeError, NameError, ImportError): # conversion to Singular not supported
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_ideal.py]: 2 spaces before `#`

File: src/sage/rings/polynomial/multi_polynomial_ideal.py [hunk 9]
```diff
@@ -4646,10 +4650,10 @@ def groebner_basis(self, algorithm='', deg_bound=None, mult_bound=None, prot=Fal
...
                 try:
                     gb = self._groebner_basis_singular("groebner", deg_bound=deg_bound, mult_bound=mult_bound, *args, **kwds)
-                except (TypeError, NameError, NotImplementedError): # conversion to Singular not supported
+                except (TypeError, NameError, NotImplementedError, ImportError): # conversion to Singular not supported
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_ideal.py]: 2 spaces before `#`

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 10]
```diff
@@ -403,17 +397,20 @@ cdef class MPolynomialRing_libsingular(MPolynomialRing_base):
 
             sage: import gc
             sage: from sage.rings.polynomial.multi_polynomial_libsingular import MPolynomialRing_libsingular
-            sage: R1 = MPolynomialRing_libsingular(GF(5), 2, ('x', 'y'), TermOrder('degrevlex', 2))
-            sage: R2 = MPolynomialRing_libsingular(GF(11), 2, ('x', 'y'), TermOrder('degrevlex', 2))
-            sage: R3 = MPolynomialRing_libsingular(GF(13), 2, ('x', 'y'), TermOrder('degrevlex', 2))
+            sage: R1 = MPolynomialRing_libsingular(GF(5), 2, ('x', 'y'),                # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: block scope

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 11]
```diff
@@ -429,22 +426,23 @@ cdef class MPolynomialRing_libsingular(MPolynomialRing_base):
...
+            sage: n = len(ring_refcount_dict)                                           # needs sage.rings.finite_rings sage.rings.function_field
+            sage: R = MPolynomialRing_libsingular(GF(547), 2, ('x', 'y'),               # needs sage.rings.finite_rings
+            ....:                                 TermOrder('degrevlex', 2))
+            sage: len(ring_refcount_dict) == n + 1                                      # needs sage.rings.finite_rings sage.rings.function_field
             True
 
-            sage: Q = copy(R)   # indirect doctest
-            sage: p = R.gen(0) ^2+R.gen(1)^2
-            sage: q = copy(p)
+            sage: Q = copy(R)   # indirect doctest                                      # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: block scope

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 12]
```diff
@@ -4269,26 +4275,27 @@ cdef class MPolynomial_libsingular(MPolynomial_libsingular_base):
...
             x^2*z^2 + x*y*z - y^2
 
         This checks that :trac:`11838` is fixed::
 
-            sage: K = GF(4,'a')
-            sage: a = K.gens()[0]
-            sage: R.<x,y> = K[]
-            sage: p=x^8*y^3 + x^2*y^9 + a*x^9 + a*x*y^4
-            sage: q=y^11 + (a)*y^10 + (a + 1)*x*y^3
+            sage: K = GF(4,'a')                                                         # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: block scope

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 13]
```diff
@@ -5375,16 +5387,16 @@ cdef class MPolynomial_libsingular(MPolynomial_libsingular_base):
...
             3*y^3 + 9*y^2*z + 9*y*z^2 + 3*z^3 - 18*y^2 - 36*y*z - 18*z^2 + 35*y + 36*z - 24
 
         Resultants are also supported over the Integers::
 
-            sage: R.<x,y,a,b,u>=PolynomialRing(ZZ, 5, order='lex')
-            sage: r = (x^4*y^2+x^2*y-y).resultant(x*y-y*a-x*b+a*b+u,x)
+            sage: R.<x,y,a,b,u> = PolynomialRing(ZZ, 5, order='lex')
+            sage: r = (x^4*y^2 + x^2*y - y).resultant(x*y - y*a - x*b + a*b + u, x)
             sage: r
             y^6*a^4 - 4*y^5*a^4*b - 4*y^5*a^3*u + y^5*a^2 - y^5 + 6*y^4*a^4*b^2 + 12*y^4*a^3*b*u - 4*y^4*a^2*b + 6*y^4*a^2*u^2 - 2*y^4*a*u + 4*y^4*b - 4*y^3*a^4*b^3 - 12*y^3*a^3*b^2*u + 6*y^3*a^2*b^2 - 12*y^3*a^2*b*u^2 + 6*y^3*a*b*u - 4*y^3*a*u^3 - 6*y^3*b^2 + y^3*u^2 + y^2*a^4*b^4 + 4*y^2*a^3*b^3*u - 4*y^2*a^2*b^3 + 6*y^2*a^2*b^2*u^2 - 6*y^2*a*b^2*u + 4*y^2*a*b*u^3 + 4*y^2*b^3 - 2*y^2*b*u^2 + y^2*u^4 + y*a^2*b^4 + 2*y*a*b^3*u - y*b^4 + y*b^2*u^2
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx:5405]: may be too much for this ticket, but the alignement could be changed

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 14]
```diff
@@ -5575,11 +5584,11 @@ cdef class MPolynomial_libsingular(MPolynomial_libsingular_base):
         ::
 
             sage: R.<x> = QQ[]
-            sage: K.<k> = NumberField(x^2 - 5)
-            sage: T.<t,w> = K[]
-            sage: I = K.ideal(3)
-            sage: f = 1/3*t*w + 3
-            sage: f.local_height(I)
+            sage: K.<k> = NumberField(x^2 - 5)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: block scope

File: src/sage/rings/polynomial/multi_polynomial_ring_base.pyx [hunk 15]
```diff
@@ -706,19 +705,20 @@ cdef class MPolynomialRing_base(sage.rings.ring.CommutativeRing):
...
     def _is_valid_homomorphism_(self, codomain, im_gens, base_map=None):
         """
         EXAMPLES::
 
             sage: T.<t> = ZZ[]
-            sage: K.<i> = NumberField(t^2 + 1)                                                              # optional - sage.rings.number_field
-            sage: R.<x,y> = K[]                                                                             # optional - sage.rings.number_field
-            sage: Q5 = Qp(5); i5 = Q5(-1).sqrt()                                                            # optional - sage.rings.number_field
-            sage: R._is_valid_homomorphism_(Q5, [Q5.teichmuller(2), Q5(6).log()]) # no coercion             # optional - sage.rings.number_field
+            sage: K.<i> = NumberField(t^2 + 1)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_ring_base.pyx]: block scope

File: src/sage/rings/polynomial/multi_polynomial_sequence.py [hunk 16]
```diff
@@ -1171,12 +1189,12 @@ def is_groebner(self, singular=singular):
 
         EXAMPLES::
 
-            sage: R.<a,b,c,d,e,f,g,h,i,j> = PolynomialRing(GF(127), 10)                 # optional - sage.rings.finite_rings
-            sage: I = sage.rings.ideal.Cyclic(R, 4)                                     # optional - sage.rings.finite_rings
-            sage: I.basis.is_groebner()                                                 # optional - sage.rings.finite_rings
+            sage: R.<a,b,c,d,e,f,g,h,i,j> = PolynomialRing(GF(127), 10)
+            sage: I = sage.rings.ideal.Cyclic(R, 4)                                     # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial_sequence.py]: block scope

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 17]
```diff
@@ -310,10 +313,10 @@ cdef class OrePolynomial(AlgebraElement):
         EXAMPLES::
 
             sage: k.<t> = GF(5^3)
-            sage: Frob = k.frobenius_endomorphism()                                     # optional - sage.rings.finite_rings
-            sage: S.<x> = k['x',Frob]                                                   # optional - sage.rings.finite_rings
-            sage: a = x + t                                                             # optional - sage.rings.finite_rings
-            sage: a[1] = t + 1                                                          # optional - sage.rings.finite_rings
+            sage: Frob = k.frobenius_endomorphism()                                     # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]: block scope

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 18]
```diff
@@ -3033,15 +3059,15 @@ cdef class OrePolynomialBaseringInjection(Morphism):
         TESTS::
 
             sage: from sage.rings.polynomial.ore_polynomial_element import OrePolynomialBaseringInjection
-            sage: k.<t> = GF(5^3)                                                       # optional - sage.rings.finite_rings
-            sage: Frob = k.frobenius_endomorphism()                                     # optional - sage.rings.finite_rings
-            sage: S.<x> = k['x',Frob]                                                   # optional - sage.rings.finite_rings
-            sage: OrePolynomialBaseringInjection(k, k['x', Frob])                       # optional - sage.rings.finite_rings
+            sage: k.<t> = GF(5^3)                                                       # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]: block scope

File: src/sage/rings/polynomial/polynomial_element.pyx [hunk 19]
```diff
@@ -1558,22 +1570,22 @@ cdef class Polynomial(CommutativePolynomial):
 
             sage: R.<x> = RDF[]
             sage: epsilon = RDF(1).ulp()*50   # Allow an error of up to 50 ulp
-            sage: f = inverse_mod(x^2 + 1, x^5 + x + 1); f  # abs tol 1e-14             # optional - sage.modules
+            sage: f = inverse_mod(x^2 + 1, x^5 + x + 1); f  # abs tol 1e-14             # needs sage.modules
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element.pyx]: block scope

File: src/sage/rings/polynomial/polynomial_element.pyx [hunk 20]
```diff
@@ -4241,170 +4290,176 @@ cdef class Polynomial(CommutativePolynomial):
...
-            sage: pol.factor()                                                          # optional - sage.rings.number_field
+            sage: pol = (-1/7*a^5 - 1/7*a^4 - 1/7*a^3 - 1/7*a^2 - 2/7*a - 1/7)*t^10 + (4/7*a^5 - 2/7*a^4 - 2/7*a^3 - 2/7*a^2 - 2/7*a - 6/7)*t^9 + (90/49*a^5 + 152/49*a^4 + 18/49*a^3 + 24/49*a^2 + 30/49*a + 36/49)*t^8 + (-10/49*a^5 + 10/7*a^4 + 198/49*a^3 - 102/49*a^2 - 60/49*a - 26/49)*t^7 + (40/49*a^5 + 45/49*a^4 + 60/49*a^3 + 277/49*a^2 - 204/49*a - 78/49)*t^6 + (90/49*a^5 + 110/49*a^4 + 2*a^3 + 80/49*a^2 + 46/7*a - 30/7)*t^5 + (30/7*a^5 + 260/49*a^4 + 250/49*a^3 + 232/49*a^2 + 32/7*a + 8)*t^4 + (-184/49*a^5 - 58/49*a^4 - 52/49*a^3 - 66/49*a^2 - 72/49*a - 72/49)*t^3 + (18/49*a^5 - 32/49*a^4 + 10/49*a^3 + 4/49*a^2)*t^2 + (2/49*a^4 - 4/49*a^3 + 2/49*a^2)*t  # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element.pyx]: may be too much for this ticket, but the alignment could be improved

File: src/sage/rings/polynomial/polynomial_element.pyx [hunk 21]
```diff
@@ -5112,33 +5177,33 @@ cdef class Polynomial(CommutativePolynomial):
 
         TESTS::
 
-            sage: R.<x> = GF(2)['x']                                                    # optional - sage.rings.finite_rings
-            sage: f = x^4 + x^3 + x^2 + x + 1                                           # optional - sage.rings.finite_rings
-            sage: f.is_primitive(15)                                                    # optional - sage.rings.finite_rings
+            sage: R.<x> = GF(2)['x']
+            sage: f = x^4 + x^3 + x^2 + x + 1
+            sage: f.is_primitive(15)                                                    # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element.pyx]: block scope

File: src/sage/rings/polynomial/polynomial_element_generic.py [hunk 22]
```diff
@@ -419,39 +425,39 @@ def __getitem__(self, n):
         EXAMPLES::
 
             sage: R.<w> = PolynomialRing(RDF, sparse=True)
-            sage: e = RDF(e)                                                            # optional - sage.symbolic
-            sage: f = sum(e^n*w^n for n in range(4)); f   # abs tol 1.1e-14             # optional - sage.symbolic
+            sage: e = RDF(e)                                                            # needs sage.symbolic
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element_generic.py]: block scope

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 23]
```diff
@@ -453,18 +457,18 @@ def _element_constructor_(self, x):
         EXAMPLES::
 
             sage: R.<x> = PolynomialRing(QQ)
-            sage: S.<alpha> = R.quotient(x^3 - 3*x + 1)                                 # optional - sage.libs.pari
-            sage: S(x)                                                                  # optional - sage.libs.pari
+            sage: S.<alpha> = R.quotient(x^3 - 3*x + 1)                                 # needs sage.libs.pari
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: block scope

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 24]
```diff
@@ -473,54 +477,56 @@ def _element_constructor_(self, x):
...
             xbar
 
         Here we test against several issues discussed in :trac:`8992`::
 
             sage: P.<x> = QQ[]
-            sage: Q1 = P.quo([(x^2+1)^2*(x^2-3)])                                       # optional - sage.libs.pari
-            sage: Q2 = P.quo([(x^2+1)^2*(x^5+3)])                                       # optional - sage.libs.pari
-            sage: p = Q1.gen() + Q2.gen()                                               # optional - sage.libs.pari
-            sage: p                                                                     # optional - sage.libs.pari
+            sage: Q1 = P.quo([(x^2+1)^2*(x^2-3)])                                       # needs sage.libs.pari
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: block scope

File: src/sage/rings/polynomial/polynomial_quotient_ring_element.py [hunk 25]
```diff
@@ -162,15 +162,15 @@ def _im_gens_(self, codomain, im_gens, base_map=None):
...
 
             sage: Zx.<x> = ZZ[]
-            sage: K.<i> = NumberField(x^2 + 1)                                          # optional - sage.rings.number_field
-            sage: cc = K.hom([-i])                                                      # optional - sage.rings.number_field
-            sage: S.<y> = K[]                                                           # optional - sage.rings.number_field
-            sage: Q.<q> = S.quotient(y^2*(y-1)*(y-i))                                   # optional - sage.rings.number_field
-            sage: T.<t> = S.quotient(y*(y+1))                                           # optional - sage.rings.number_field
-            sage: phi = Q.hom([t+1], base_map=cc)                                       # optional - sage.rings.number_field
-            sage: phi(q)                                                                # optional - sage.rings.number_field
+            sage: K.<i> = NumberField(x^2 + 1)                                          # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring_element.py]: block scope

File: src/sage/rings/polynomial/polynomial_quotient_ring_element.py [hunk 26]
```diff
@@ -470,50 +470,51 @@ def field_extension(self, names):
...
 
         Another more awkward example::
 
             sage: R.<x> = QQ['x']
-            sage: K.<a> = NumberField(x^3 - 2)                                          # optional - sage.rings.number_field
-            sage: S.<X> = K['X']                                                        # optional - sage.rings.number_field
-            sage: f = (X+a)^3 + 2*(X+a) + 1                                             # optional - sage.rings.number_field
-            sage: f                                                                     # optional - sage.rings.number_field
+            sage: K.<a> = NumberField(x^3 - 2)                                          # needs sage.rings.number_field
+            sage: S.<X> = K['X']                                                        # needs sage.rings.number_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring_element.py]: block scope

File: src/sage/rings/rational_field.py [hunk 27]
```diff
@@ -657,7 +657,7 @@ def places(self, all_complex=False, prec=None):
 
         EXAMPLES::
 
-            sage: QQ.places()
+            sage: QQ.places()                                                           # optional - sage.rings.real_mpfr
```
Comment (dcoudert) [inline src/sage/rings/rational_field.py]: `# needs`

**Review by dcoudert — CHANGES_REQUESTED** [review]
This is a giant ticket. Can't it be split per sub directories ?

There are several files in which you used `# optional` and not `# needs`. Also, block scope could be employed in many places. I noted only a few of these places (too many).

**Comment by mkoeppe** [comment]
> This is a giant ticket. Can't it be split per sub directories ?

There were a few commits with incidental changes in other directories. I have trimmed them away now.

> There are several files in which you used `# optional` and not `# needs`. Also, block scope could be employed in many places. I noted only a few of these places (too many).

Thanks for your patient review of this megaticket. I've made the indicated changes and many more.

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 28]
```diff
@@ -1377,23 +1396,23 @@ cdef class OrePolynomial(AlgebraElement):
...
-            sage: D = S.random_element(degree=2)                                        # optional - sage.rings.finite_rings
-            sage: P = D * S.random_element(degree=2)                                    # optional - sage.rings.finite_rings
-            sage: Q = D * S.random_element(degree=2)                                    # optional - sage.rings.finite_rings
-            sage: L = P.right_lcm(Q)                                                    # optional - sage.rings.finite_rings
-            sage: U = right_lcm_cofactor(P, Q)                                          # optional - sage.misc.cython sage.rings.finite_rings
-            sage: (P*U).left_monic() == L                                               # optional - sage.misc.cython sage.rings.finite_rings
+            sage: k.<a> = GF(7^5)
+            sage: Frob = k.frobenius_endomorphism(3)
+            sage: S.<x> = k['x', Frob]
+            sage: # needs sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]: This is redundant with line 1399

File: src/sage/rings/polynomial/ore_polynomial_element.pyx [hunk 29]
```diff
@@ -1377,23 +1396,23 @@ cdef class OrePolynomial(AlgebraElement):
...
+            sage: D = S.random_element(degree=2)                                        # needs sage.misc.cython
```
Comment (dcoudert) [inline src/sage/rings/polynomial/ore_polynomial_element.pyx]: the `# needs` is redundant with line 1399

File: src/sage/rings/polynomial/polynomial_element.pyx [hunk 30]
```diff
@@ -10040,83 +10243,84 @@ cdef class Polynomial(CommutativePolynomial):
...
 
-            sage: all(cyclotomic_polynomial(i).is_cyclotomic() for i in range(1, 101))  # optional - sage.libs.pari
+            sage: all(cyclotomic_polynomial(i).is_cyclotomic() for i in range(1, 101))  # needs sage.libs.pari
             True
 
         Some more tests::
 
-            sage: f = x^16 + x^14 - x^10 + x^8 - x^6 + x^2 + 1
-            sage: f.is_cyclotomic(algorithm="pari")                                     # optional - sage.libs.pari
+            saeg: # needs sage.libs.pari
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element.pyx]: `saeg` -> `sage` ;)

File: src/sage/rings/polynomial/polynomial_element.pyx [hunk 31]
```diff
@@ -10040,83 +10243,84 @@ cdef class Polynomial(CommutativePolynomial):
...
+            sage: f = x^16 + x^14 - x^10 + x^8 - x^6 + x^2 + 1                          # needs sage.libs.pari
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_element.pyx]: `# needs` redundant with above line

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 32]
```diff
@@ -1,22 +1,24 @@
-# -*- coding: utf-8 -*-
+# sage.doctest: needs sage.libs.pari
 """
 Quotients of Univariate Polynomial Rings
 
 EXAMPLES::
 
+    sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: to remove

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 33]
```diff
@@ -79,27 +80,28 @@ class PolynomialQuotientRingFactory(UniqueFactory):
     We create the quotient ring `\ZZ[x]/(x^3+7)`, and
     demonstrate many basic functions with it::
 
+        sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: to be removed

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 34]
```diff
@@ -128,23 +131,23 @@ class PolynomialQuotientRingFactory(UniqueFactory):
     Next we create a number field, but viewed as a quotient of a
     polynomial ring over `\QQ`::
 
+        sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: again

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 35]
```diff
@@ -195,19 +198,20 @@ def create_key(self, ring, polynomial, names=None):
 
         Consequently, you get two distinct objects::
 
-            sage: S = PolynomialQuotientRing(R, x + 1); S                               # optional - sage.libs.pari
+            sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: again

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 36]
```diff
@@ -330,24 +334,25 @@ class of the quotient ring and its newly created elements.
...
         ['cartesian_product', 'inverse', 'inverse_of_unit', 'is_idempotent',
          'is_one', 'is_unit', 'lift', 'powers']
-        sage: first_class = Q.__class__                                                 # optional - sage.libs.pari
+        sage: first_class = Q.__class__
 
     We try to find out whether `Q` is a field. Indeed it is, and thus its category,
     including its class and element class, is changed accordingly::
 
-        sage: Q in Fields()                                                             # optional - sage.libs.pari
+        sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: another

File: src/sage/rings/polynomial/polynomial_quotient_ring.py [hunk 37]
```diff
@@ -473,54 +478,56 @@ def _element_constructor_(self, x):
...
-            sage: p.parent()('xbar')                                                    # optional - sage.libs.pari
+            sage: p.parent()('xbar')
             xbar
 
         Note that the result of string conversion has the correct parent, even
         when the given string suggests an element of the cover ring or the base
         ring::
 
-            sage: a = Q1('x'); a                                                        # optional - sage.libs.pari
+            sage:
```
Comment (dcoudert) [inline src/sage/rings/polynomial/polynomial_quotient_ring.py]: a few other places in this file. Result of a script/command.

File: src/sage/rings/polynomial/term_order.py [hunk 38]
```diff
@@ -1822,12 +1827,12 @@ def macaulay2_str(self):
 
         EXAMPLES::
 
-            sage: P = PolynomialRing(GF(127), 8, names='x',                             # optional - sage.rings.finite_rings
+            sage: P = PolynomialRing(GF(127), 8, names='x',
```
Comment (dcoudert) [inline src/sage/rings/polynomial/term_order.py]: single line ?

File: src/sage/rings/polynomial/laurent_polynomial_ring_base.py [hunk 39]
```diff
@@ -391,14 +392,14 @@ def _is_valid_homomorphism_(self, codomain, im_gens, base_map=None):
...
-            sage: L._is_valid_homomorphism_(K, (K(1/2), K(3/2)))                                    # optional - sage.rings.number_field
+            sage: K.<i> = NumberField(t^2 + 1)                                                      # needs sage.rings.number_field
+            sage: L.<x,y> = LaurentPolynomialRing(K)                                                # needs sage.rings.number_field
+            sage: L._is_valid_homomorphism_(K, (K(1/2), K(3/2)))                                    # needs sage.rings.number_field
             True
-            sage: Q5 = Qp(5); i5 = Q5(-1).sqrt()                                                    # optional - sage.rings.padics
-            sage: L._is_valid_homomorphism_(Q5, (Q5(1/2), Q5(3/2))) # no coercion                   # optional - sage.rings.padics
+            sage: Q5 = Qp(5); i5 = Q5(-1).sqrt()                                                    # needs sage.rings.padics
+            sage: L._is_valid_homomorphism_(Q5, (Q5(1/2), Q5(3/2))) # no coercion                   # needs sage.rings.number_field sage.rings.padics
```
Comment (dcoudert) [inline src/sage/rings/polynomial/laurent_polynomial_ring_base.py]: 2 spaces before `#`

File: src/sage/rings/polynomial/multi_polynomial.pyx [hunk 40]
```diff
@@ -44,6 +40,7 @@ cdef class MPolynomial(CommutativePolynomial):
         r"""
         TESTS::
 
+            sage: # needs sage.rings.real_mpfr
             sage: ZZ(RR['x,y'](0)) # indirect doctest
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial.pyx]: 2 spaces before `#`

File: src/sage/rings/polynomial/multi_polynomial.pyx [hunk 41]
```diff
@@ -2419,7 +2438,7 @@ cdef class MPolynomial(CommutativePolynomial):
             sage: R.<x,y> = PolynomialRing(RR)
             sage: F = (217.992172373276*x^3 + 96023.1505442490*x^2*y
             ....:      + 1.40987971253579e7*x*y^2 + 6.90016027113216e8*y^3)
-            sage: F.reduced_form(smallest_coeffs=False) # tol 1e-8                      # optional - sage.modules
+            sage: F.reduced_form(smallest_coeffs=False) # tol 1e-8                      # needs sage.modules sage.rings.complex_interval_field
```
Comment (dcoudert) [inline src/sage/rings/polynomial/multi_polynomial.pyx]: 2 spaces before `# tol` ?

**Review by dcoudert — CHANGES_REQUESTED** [review]
This is much better, but I noted a few more issues.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Pull request merged**